packages/react-dev-utils/failFast.js

```diff
@@ -0,0 +1,394 @@
+  // Revisit Jan 2016
+  // https://developer.mozilla.org/en-US/Firefox/Releases/51#JavaScript
+  // https://bugzilla.mozilla.org/show_bug.cgi?id=1101653
+  let omittedFramesCount = 0
+  const appendOmittedFrames = () => {
+    if (!omittedFramesCount) return
+    const omittedFrames = document.createElement('div')
+    omittedFrames.appendChild(document.createTextNode(`---[ ${omittedFramesCount} internal calls hidden ]---`))
+    applyStyles(omittedFrames, omittedFramesStyle)
+    trace.appendChild(omittedFrames)
+    omittedFramesCount = 0
+  }
+  let main = true
+  for (let frame of resolvedFrames) {
+    const {
+      functionName,
+      fileName, lineNumber, columnNumber,
+      scriptLines,
+      sourceFileName, sourceLineNumber, sourceColumnNumber,
+      sourceLines
+    } = frame
+
+    // Skip native functions like Array.forEach
+    if (fileName === '(native)') continue;
+
+    let url
+    if (sourceFileName) {
+      url = sourceFileName + ':' + sourceLineNumber
+      if (sourceColumnNumber) url += ':' + sourceColumnNumber
+    } else {
+      url = fileName + ':' + lineNumber
+      if (columnNumber) url += ':' + columnNumber
+    }
+
+    const internalUrl = isInternalFile(url)
+    if (internalUrl && internalDisabled) {
+      omittedFramesCount++
+      continue
+    }
+
+    appendOmittedFrames()
+
+    const elem = frameDiv(functionName, url, internalUrl);
+
+    if (!internalUrl) {
+      if (sourceDisabled && scriptLines.length !== 0) {
+        elem.appendChild(sourceCodePre(scriptLines, lineNumber, columnNumber, main))
+      } else if (!sourceDisabled && sourceLines.length !== 0) {
+        elem.appendChild(sourceCodePre(sourceLines, sourceLineNumber, sourceColumnNumber, main))
+      }
+      main = false
+    }
+
+    trace.appendChild(elem)
+  }
+  appendOmittedFrames()
+  return trace;
+}
+
+function render(error, name, message, resolvedFrames) {
+  if (overlayReference !== null) {
+    renderAdditional()
+    return
+  }
+
+  // Create overlay
+  const overlay = document.createElement('div')
+  applyStyles(overlay, overlayStyle)
+  overlay.appendChild(hintsDiv())
+
+  // Create container
+  const container = document.createElement('div')
+  applyStyles(container, containerStyle)
+  overlay.appendChild(container)
+
+  // Create header
+  const header = document.createElement('div')
+  applyStyles(header, headerStyle)
+  header.appendChild(document.createTextNode(`${name}: ${message}`))
+  container.appendChild(header)
+
+  // Create trace
+  container.appendChild(traceDiv(resolvedFrames))
+
+  // Mount
+  document.body.appendChild(overlayReference = overlay)
+  errorCache = error
+  additionalCount = 0
+}
+
+function unmount() {
+  if (overlayReference === null) return
+  document.body.removeChild(overlayReference)
+  overlayReference = null
+}
+
+function isInternalFile(url) {
+  return url.indexOf('/~/') !== -1 || url.trim().indexOf(' ') !== -1
+}
+
+function crash(error, unhandledRejection = false) {
```

Comment:
@gaearon what do you think about doing `if (module.hot) module.hot.decline()` on first sign of error so that the page **has to be reloaded**? (I know we don't do hot reloading yet, but future proofing ...)